From 765e628cb2ad2153cf6adc581262da7abd0d9213 Mon Sep 17 00:00:00 2001
From: Wim Taymans <wtaymans@redhat.com>
Date: Mon, 5 Oct 2026 11:16:25 +0200
Subject: [PATCH] gstpipewiresrc: fix locking order

The locking order is thread_lock > object_lock

In buffer recycle take the thread lock before the object lock.

The dispose function is set in add_buffer and removed in remove_buffer,
which have the thread_lock taken. Check this only once and as soon as the
thread_lock is acquired.

Fixes #5507

diff --git a/src/gst/gstpipewiresrc.c b/src/gst/gstpipewiresrc.c
index fb48a0e65..ade63758b 100644
--- a/src/gst/gstpipewiresrc.c
+++ b/src/gst/gstpipewiresrc.c
@@ -581,21 +581,15 @@ buffer_recycle (GstMiniObject *obj)
     return TRUE;
   }
 
-  GST_OBJECT_LOCK (data->pool);
-  if (!obj->dispose) {
-    GST_OBJECT_UNLOCK (data->pool);
-    return TRUE;
-  }
-
-  GST_BUFFER_FLAGS (obj) = data->flags;
-
   pw_thread_loop_lock (src->stream->core->loop);
   if (!obj->dispose) {
     pw_thread_loop_unlock (src->stream->core->loop);
-    GST_OBJECT_UNLOCK (data->pool);
     return TRUE;
   }
 
+  GST_OBJECT_LOCK (data->pool);
+  GST_BUFFER_FLAGS (obj) = data->flags;
+
   gst_mini_object_ref (obj);
 
   data->queued = TRUE;
@@ -607,9 +601,8 @@ buffer_recycle (GstMiniObject *obj)
   else
     GST_LOG_OBJECT (src, "recycle buffer %p", obj);
 
-  pw_thread_loop_unlock (src->stream->core->loop);
-
   GST_OBJECT_UNLOCK (data->pool);
+  pw_thread_loop_unlock (src->stream->core->loop);
 
   return FALSE;
 }

